Gentle Buffer Stock improvements

sbenthall opened the pull request:
Work on Gentle-Intro-To-HARK-Buffer-Stock-Model
Issue #87

sbenthall commented:
@llorracc This PR is ready for your review.

I did change some things about parameters and inheritance.

Two things I haven't been able to move forward on:
 - The parameter inheritance does not yet make the most of the class inheritance between AgentTypes; I hope to do some more work on this. So what I included in this notebook is agnostic to the implementation of the parameter inheritance.
 - I am still struggling to understand exactly how the universal solver and its instantiations works. So instead of trying to explain it, I've introduced commands that allow the students to inspect the documentation themselves.

sbenthall commented:
@llorracc This might be easier for you to review: https://github.com/sbenthall/DemARK/blob/i87/notebooks/Gentle-Intro-To-HARK-Buffer-Stock-Model.ipynb

llorracc commented:
Thx - but what’s the diff btw or and link below?

On Mon, Feb 17, 2020 at 00:06 Sebastian Benthall <notifications@github.com>
wrote:

> @llorracc <https://github.com/llorracc> This might be easier for you to
> review:
> https://github.com/sbenthall/DemARK/blob/i87/notebooks/Gentle-Intro-To-HARK-Buffer-Stock-Model.ipynb
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/econ-ark/DemARK/pull/88?email_source=notifications&email_token=AAKCK742VB3LAPCDBK3BJ43RDHBJDA5CNFSM4KUXQUC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEL4VKVY#issuecomment-586765655>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKCK77LXSRHW6OBSTLUR2TRDHBJDANCNFSM4KUXQUCQ>
> .
>
-- 
Sent from Gmail Mobile

sbenthall commented:
You can view the diff here: https://github.com/econ-ark/DemARK/pull/88/files

- reduced the explicit parameters set in code to the necessary ones for the content
- added sections about using `dir()` and `__bases__` to better understand inheritance
- added a visualization of the discretized shock distribution
- added a section on the `help()` function in Python

llorracc commented:
Sorry, I was unclear: Was asking what the difference was between the version you gave me a link to, and the one that is in the PR?

PS.  When I test a "merge" I get a message that the branch has conflicts that must be resolved.

sbenthall commented:
The one I linked to and the one in the PR are identical.
The link and the PR point in different ways to the same repository, branch, and file.

I've fixed the merge conflicts just now.

llorracc commented:
OK, got it, thanks.

Am just looking at it now.

1. The "setup" cell still has %matplotinline.  
    * I believe that this is no longer necessary, right?
1. I like what you have done.  
1. But would like to expand the content a bit
    * In particular, you are more of a newcomer to HARK than I am
    * So the things that students might stumble over and need explained are fresher in your mind than mine
1. You've spent some time reviewing both Karsten's notebook and Jan Rosa's.  
    * Let's think of this notebook as prep material for one of those 
    * If you look at those with an eye toward "what things might be good prep for these notebooks" am guessing you can quickly come up with a couple of things in addition to the point about how you have to reconstruct shocks.

sbenthall commented:
Ok, in the most recent commits I've:
- removed the `%matplotlib` magic
- added a section on `backwards induction`, which is an opportunity to discuss the Solution object and the functions within it, including an introduction of the linear interpolator used for the consumption function.

This sets up students for the discussion of the solver mechanics in the next lecture.

In general, I'd say the interpolation mechanics are one of the bigger surprises to me in the library; it's also clearly substantively important functionality. I'd like to surface it more in the documentation elsewhere as well.
